### DISCUSSION TITLE
"Failed to fetch dynamically imported module" error in admin console

### DISCUSSION THREAD

**Discussion by Nachwahl** [keycloak/keycloak#22669] [Q&A] [answered]
Hi there.

I recently updated to the newest Keycloak version. Since that update, I'm not able to access any client settings anymore but get this error: `Failed to fetch dynamically imported module: https://my.keycloak.instance/resources/v4npb/admin/keycloak.v2/assets/ClientDetails-3ac5360d.js` (see screenshot)
![image](https://github.com/keycloak/keycloak/assets/43521040/e2cfd977-bfb3-4eb0-ac88-abe36bfc5090)

Looking at the browser console, it seems like it's getting an 404 error when loading those scripts, while other scripts (such as `https://my.keycloak.instance/resources/v4npb/admin/keycloak.v2/assets/DraggableTable-88759caa.js`) load perfectly fine. What could cause that problem?

**Comment by ssilvert** [comment]
Maybe try clearing your browser cache?

**Reply by Nachwahl** [reply]
I already tried clearing the cache and using another browser :/

**Reply by Nachwahl** [reply]
I also noticed that the "Welcome to Keycloak" page looks off. It seems like the welcome.css file also returns a 404 error code.

**Comment by NagenderPulluri** [comment]
I'm also facing the same error often, for temporary resolution I'm restarting the pods, but getting the same error after some time, couldn't able to find what is the issue, logs not showing anything related this error. After upgrading keycloak from 18.0.2 to 21.1.1 version, i'm getting this error.

**Comment by NagenderPulluri** [comment]
@Nachwahl any luck ?

**Comment by jonkoops** [comment]
If someone has a reproducible case of this issue I would recommend that you log an issue with detailed steps on how to reproduce it.

**Comment by tobiwanobi** [comment]
Same issue with v22, downgraded to 21.1.2 and works perfectly fine. Faced this issue only when using a HA deployment.

**Reply by tobiwanobi** [reply]
I upgraded back to 22.0.4 and now it looks like everything is working again...

**Comment by basebox-mt** [comment, marked as answer]
Seems to be a permission problem. What fixed it for me (assuming keycloak being installed in `/opt/keycloak`):

```sh
chown -R keycloak /opt/keycloak
cd /opt/keycloak
# Rebuild as keycloak user
sudo -u keycloak bin/kc.sh build
```

**Reply by pengfei99** [reply]
I confirm this solution works if you built your keycloak with root user and try to run it with keycloak user.

**Reply by codespearhead** [reply]
Without a [minimal reproducible example](https://en.wikipedia.org/wiki/Minimal_reproducible_example#:~:text=In%20computing%2C%20a%20minimal%20reproducible,to%20be%20demonstrated%20and%20reproduced.) as suggested by Jon Koops ( @jonkoops ) or at least more information on your environment, such as the console logs in your browser's web development tools, it might be impossible to address this issue.

However, the solution attempt mentioned above makes the most sense to me.

Since **it's been about a year without feedback from you ( @Nachwahl )** , can you at least close this discussion so it doesn't show up in front of discussions yet to be solved?

I wonder if Jon has permissions to do so.

**Comment by dilipk8** [comment]
Is there any permanent fix for this issue ? I get the same error.

**Reply by basebox-mt** [reply]
"Permanent fix" is to run build as the same user that later runs keycloak.

**Comment by Piosie** [comment]
In my case respond @basebox-mt is correct. I migrated Keycloak 18.0.2 to 22.0.1 and I got the same problem. after change permisson on 'keycloak/data' folder to owner 'keycloak' I don't have problems.

**Comment by EML-github** [comment]
I had a very similar error on a new 23.0.1 install (a permissions error on fetching a dynamically imported JS file). The answer from @basebox-mt leaves the whole `/opt/keycloak` tree owned by user `keycloak`, which is a security issue. My procedure is:

1. Lock down the install with `chown -R root:root /opt/keycloak`
2. Open up keycloak/data with `chown -R keycloak:keycloak /opt/keycloak/data`
3. Run the build as user `root` (I haven't tried building as user `keycloak`; if that works, great, do that instead)
4. Run keycloak as user `keycloak`

This works. My problem in this case was that I had only set the `/opt/keycloak/data` directory itself to owner `keycloak:keycloak` (I missed out the `-R` in step 2 above). Repeating that step fixed everything.

**Comment by Freddyp91** [comment]
I actually run into this issue every so often after a upgrade..

For me keycloak is deployed to a k8s cluster. I just restart the pod and it works with no issue.
